Cardinality

vpetrovykh opened the pull request:
We can squash the two commits into one. I just wanted to keep purely renaming one separate to make it easier to see what's going on.

Regarding naming: 
- The `SchemaCardinality` is what gets written to or read from the schema. So the field name should match the schema field "cardinality".
- Ideally the inferred cardinality with upper and lower bounds should have a different field/attribute name to reduce confusion. I'm open to suggestions for alternatives to "full_cardinality".

Regarding `CardinalityBound`: the bounds are used in conversion from schema "requirement + cardinality" format to a single enum and a some calculations. This makes it hard to refactor the class into `cardinality.py` cleanly, since then the conversion is invoked in a few places outside it.

vpetrovykh commented on edb/edgeql/compiler/__init__.py:
```diff
@@ -519,7 +519,7 @@ def compile_func_to_ir(
     return_typemod = func.get_return_typemod(schema)
     if (return_typemod is not qltypes.TypeModifier.SET_OF
-            and ir.cardinality is qltypes.SchemaCardinality.MANY):
+            and ir.full_cardinality.is_multi()):
```
Is there a reason why you want `cardinality` here? I'd rather have consistent naming for the two different cardinality enum attributes. As I explained in the PR blurb, "cardinality" is ultimately used as a property name in schema, so it makes sense to keep that name for things that map onto that, whereas any other notion of cardinality should have a clearly distinct name to avoid confusion.

elprans replied:
`full_cardinality` rubs me the wrong way, especially considering the inconsistency in the naming of the enum classes themselves.  Also, I don't think there's really that much room for confusion.  The IR and most of the compiler pipeline does not interact with schema cardinality directly, and all places in IR where you currently use `SchemaCardinality` can easily be replaced with `Cardinality`.

vpetrovykh replied:
I've renamed things back to "cardinality".

vpetrovykh commented on edb/edgeql/qltypes.py:
```diff
@@ -73,16 +76,97 @@ class TransactionDeferMode(s_enum.StrEnum):
             return 'multi'
 
 
+class CardinalityBound(int, enum.Enum):
```
Maybe I'm misunderstanding you here, but making it private to inference module (I assume you mean `edb/edgeql//compiler/inference/cardinality.py` would make it hard to use in conversion functions `from_schema_value` (used in `edb/schema/pointers.py`) and `to_schema_value` (used in `edb/schema/pointers.py` and `edb/edgeql/compiler/stmtctx.py`). Do you want me to move the conversions to `astutils.py` or?

elprans replied:
`get_full_cardinality` itself seems to be dead code.

vpetrovykh replied:
Ah, I missed that part about dead code, nice catch (I used it in some earlier iterations and forgot). I can definitely refactor it and move the inference stuff now.

elprans replied:
Also, if that piece is unused, how are you actually taking into account `required` on a pointer?  `BasePointerRef` seems to still be using schema cardinalities, you probably want to adjust `ptrref_from_ptrcls`.

vpetrovykh replied:
Code moved.

elprans commented on edb/edgeql/compiler/context.py:
```diff
@@ -111,7 +111,7 @@ def __call__(
 class PendingCardinality(NamedTuple):
 
-    specified_cardinality: Optional[qltypes.Cardinality]
+    specified_cardinality: Optional[qltypes.SchemaCardinality]
```
This place, in particular, should use `qltypes.Cardinality` instead

vpetrovykh replied:
Hmm, the `specified_cardinality` is taken from the schema in the current implementation. That's why I left it as `SchemaCardinality`. 

However, if we wanted to convert the schema cardinality to inference cardinality, that's where we'd need those helpers dealing with bounds and also later on to check if the inferred cardinality falls within the bounds of the specified one. Technically, we could get away with just the `is_single`/`is_multi` helpers, but only if we don't allow `required` as part of the spec. If I can specify that my computable is a `required multi` (it was my understanding that this is actually one of the end goals of this cardinality change in the first place), then I need proper bound checking to ensure that inferred `AT_MOST_ONE` is not acceptable as specified `AT_LEAST_ONE`, but `ONE` is. Incidentally, this would make CardinalityBound again move closer to Cardinality in terms of packaging so that the utilities to compare cardinalities can use the bounds conveniently.

elprans replied:
I don't see why `inference.cardinality` could not export a helper function to compare cardinalities.  `qltypes` is supposed to be a simple set of dumb enums to be reused across schema and various ASTs.

vpetrovykh replied:
OK, I've renamed everything back to "cardinality".

vpetrovykh commented on edb/edgeql/qltypes.py:
```diff
@@ -73,16 +76,97 @@ class TransactionDeferMode(s_enum.StrEnum):
+        return req, scard
+
+
+CARD_TO_TUPLE = {
```
Code moved.

vpetrovykh commented on edb/edgeql/qltypes.py:
```diff
@@ -73,16 +76,97 @@ class TransactionDeferMode(s_enum.StrEnum):
+}
+
+
+TUPLE_TO_CARD = {
```
Code moved.

vpetrovykh commented:
I've refactored things so that the extra classes live only in the cardinality inference module and I have renamed my "full_cardinality" back to "cardinality". At the moment these changes are spread over 3 commits:
1) Just mechanically rename old class Cardinality to SchemaCardinality.
2) Introduce new mechanism for cardinality inference and corresponding "full_cardinality" attribute to make it easier to find new usage.
3) Rename the "full_cardinality" back to "cardinality".

Once this we're happy with this I can just squash them all into one commit and merge.

elprans commented:
Getting there, thanks! 

I still don't see this resolved:

> how are you actually taking into account `required` on a pointer? `BasePointerRef` seems to still be using schema cardinalities, you probably want to adjust `ptrref_from_ptrcls`.

I.e.:  a shape computable using a required pointer must resolve to `AT_LEAST_ONE`
